Add language attribute to canvas

Aditi-1400 opened the pull request:
Fixes issue #16843.

In certain cases, the text layer was misaligned due to a difference 
between the `lang` attribute of the viewer and the canvas. This commit addresses the problem by adding the `lang` attribute to the canvas. The issue was caused because PDF.js uses serif/sans-serif fonts to generate the text layer and relies on system fonts. The difference in the `lang` attribute led to different fonts being picked, causing the misalignment.

Before the change:
<img width="783" alt="image" src="https://github.com/mozilla/pdf.js/assets/62544124/13ef8c40-a1cc-4bfa-bfe4-dc5f57a2b8c4">

After the change:
<img width="726" alt="image" src="https://github.com/mozilla/pdf.js/assets/62544124/5f82333b-4fd3-4b83-afa1-6f5d2498dec2">

Thanks, @nicolo-ribaudo for helping with this.

Snuffleupagus reviewed (changes requested):
Sorry, but this change isn't correct.
First of all the textLayer can be used outside of the viewer, a use-case that this patch completely breaks. Secondly you're not really "allowed" to do inline DOM lookups like this, hence a different solution would be required.

calixteman commented:
I'd say that the right thing to do is to attach the canvas at the right place here:
https://github.com/mozilla/pdf.js/blob/master/src/display/text_layer.js#L80

Aditi-1400 commented:
> Sorry, but this change isn't correct. First of all the textLayer can be used outside of the viewer, a use-case that this patch completely breaks. Secondly you're not really "allowed" to do inline DOM lookups like this, hence a different solution would be required.

Makes sense, um, I believe to fix this, getCtx() could be changed to set the lang attribute of the canvas to be the same as that of the viewer?

Aditi-1400 commented:
@calixteman @Snuffleupagus Hey, I updated the pull request, so we assign the same language as the viewer to the canvas, this fixes the text layer issue and also the problems that @calixteman  mentioned above.

nicolo-ribaudo commented:
I wonder what "the right place" would be. Currently the language is set per-page, while there is only one canvas injected in the document.

Are all pages of a PDF guaranteed to have the same language? If so we could inject the canvas in the first page's text layer, and re-use it for all the pages. If not, the alternative is to inject the canvas in every page.

calixteman commented:
I suppose we could pass the `viewer` to the text layer in order to attach the canvas to it but it'd almost the same amount of changes. Or we add something like `const element = container.closest("[lang]:not([lang=''])") || document.body;` when creating the canvas context (but only once the `lang` attribute has been added).
I think my preference would be to pass the lang as it's done in this patch... but usually I rely on @Snuffleupagus opinion for this kind of questions.
@Snuffleupagus, could you imagine a better/simpler solution ?

Snuffleupagus reviewed (changes requested):
Unfortunately this is *not* a good solution in its current form, since viewer initialization and first rendering *is now blocked* on Metadata fetching and parsing despite that being unrelated to actual rendering.

This will thus affect *general* viewer loading performance for all PDF documents, given that /Info-dictionaries and /Metadata-streams are often placed *at the end* of the file. Especially for e.g. linearized PDFs this could thus have a noticeable impact.

Hence my first idea, to work-around these problems, could be to also include the /Lang-data in the `textContent` such that it becomes directly available in the `src/display/text_layer.js` file. (I can *try* outlining a patch for this part during the weekend).

calixteman commented:
> Unfortunately this is _not_ a good solution in its current form, since viewer initialization and first rendering _is now blocked_ on Metadata fetching and parsing despite that being unrelated to actual rendering.
> 
> This will thus affect _general_ viewer loading performance for all PDF documents, given that /Info-dictionaries and /Metadata-streams are often placed _at the end_ of the file. Especially for e.g. linearized PDFs this could thus have a noticeable impact.
> 

That's a fair point.

> Hence my first idea, to work-around these problems, could be to also include the /Lang-data in the `textContent` such that it becomes directly available in the `src/display/text_layer.js` file. (I can _try_ outlining a patch for this part during the weekend).

Passing the `lang` in the `textContent` will still induce to fetch and parse the metadata probably too soon and then induce a less noticeable impact but still an impact (and it should slightly increase the memory use).
I suppose that having a lang isn't that frequent, if this assertion is correct, maybe we could just recompute the text layer (just the scale factors) when `lang !== ""`, else we could create a text layer without any scale factors and when the `lang` is set then compute them.
That said I should really work again on that patch I've to use the pdf fonts in the text layer because I guess that changing the `lang` attribute won't impact the fonts used in the text layer.

Aditi-1400 commented:
> Passing the `lang` in the `textContent` will still induce to fetch and parse the metadata probably too soon and then induce a less noticeable impact but still an impact (and it should slightly increase the memory use). I suppose that having a lang isn't that frequent, if this assertion is correct, maybe we could just recompute the text layer (just the scale factors) when `lang !== ""`, else we could create a text layer without any scale factors and when the `lang` is set then compute them. That said I should really work again on that patch I've to use the pdf fonts in the text layer because I guess that changing the `lang` attribute won't impact the fonts used in the text layer.

I did do some experimentation to make the text layer font the same as the document font, but I didn't go ahead with that since the font details are cleaned up after a few seconds [here](https://github.com/mozilla/pdf.js/blob/master/web/app.js#L1721) unless `fontExtraProperties` is enabled. However, if this is a preferred approach, maybe I could help with completing the patch?

calixteman commented:
> > Passing the `lang` in the `textContent` will still induce to fetch and parse the metadata probably too soon and then induce a less noticeable impact but still an impact (and it should slightly increase the memory use). I suppose that having a lang isn't that frequent, if this assertion is correct, maybe we could just recompute the text layer (just the scale factors) when `lang !== ""`, else we could create a text layer without any scale factors and when the `lang` is set then compute them. That said I should really work again on that patch I've to use the pdf fonts in the text layer because I guess that changing the `lang` attribute won't impact the fonts used in the text layer.
> 
> I did do some experimentation to make the text layer font the same as the document font, but I didn't go ahead with that since the font details are cleaned up after a few seconds [here](https://github.com/mozilla/pdf.js/blob/master/web/app.js#L1721) unless `fontExtraProperties` is enabled. However, if this is a preferred approach, I could look more into this, and help with completing the patch?

The cache thing isn't really a problem here: we could just disable it, the main problem is that we've to deal with a lot of fonts which aren't always in a good state to be used as is in the text layer.
For example some of them have no width table or have a widths table different of the one used by the pdf...  so the fonts must be amended to have the correct data.
And I had some issues with some pdf in arabic because they're using some ligatures which aren't specified in font itself.
I must rebase the patch I've and I'll share it with you.
That said, after having thought about that, in some cases (like with type3 fonts) we still need to use the text layer as it is right now.
So it's worth having a fix for this issue without refactoring the text layer stuff.

Snuffleupagus commented:
> Passing the `lang` in the `textContent` will still induce to fetch and parse the metadata probably too soon and then induce a less noticeable impact but still an impact (and it should slightly increase the memory use).

My idea would *only* require invoking https://github.com/mozilla/pdf.js/blob/e78ce74b5fd067c563354e6c118379cb4a926d09/src/core/catalog.js#L117-L124 and not until fetching the `textContent`.

calixteman commented:
Can't we just get a promise with the lang and pass it when building the text layer and get its value only when the first chunk is rendered, something like that ?

timvandermeij commented:
Now that #17941 is in place, this patch can be updated to make use of the new `lang` attribute, which should hopefully simplify this patch.

Aditi-1400 committed:
Add `lang` attribute to canvas element

Fixes issue #16843.
In certain cases, the text layer was misaligned
due to a difference between the `lang` attribute
of the viewer and the canvas. This commit addresses
the problem by adding the `lang` attribute to the canvas.

The issue was caused because PDF.js uses serif/sans-serif
fonts to generate the text layer and relies on system fonts.
The difference in the `lang` attribute led to different fonts
being picked, causing the misalignment.

Aditi-1400 commented:
> Now that #17941 is in place, this patch can be updated to make use of the new `lang` attribute, which should hopefully simplify this patch

Updated the pull request using the new `lang` attribute.  
Also, makeref is failing so I didn't run snapshot tests locally, but this shouldn't affect rendering

nicolo-ribaudo commented on src/display/text_layer.js:
```diff
@@ -65,6 +65,9 @@ const ascentCache = new Map();
 
 function getCtx(lang = null) {
+  if (_canvasContext && _canvasContext.canvas.lang !== (lang || "")) {
+    cleanupTextLayer();
```
This function should probably be renamed to `cleanupTextLayerCanvas` or similar. While reading this diff I was very confused as to why when we have to render some text with a different language we need to cleanup the whole text layer.

Snuffleupagus replied:
Unfortunately this could break things quite badly outside of the viewer, since when the library API is used directly multiple PDF documents could be active at once and this could "pull the rug" on another currently rendering textLayer from another document (and thus with another `lang` parameter).

---

Hence I believe that the "correct" solution would be to instead change `_canvasContext` to a `Map` and thus cache (potentially multiple) temporary canvases on the `lang` value.

nicolo-ribaudo commented on src/display/text_layer.js:
```diff
@@ -65,6 +65,9 @@ const ascentCache = new Map();
 let _canvasContext = null;
 
 function getCtx(lang = null) {
+  if (_canvasContext && _canvasContext.canvas.lang !== (lang || "")) {
```
```suggestion
function getCtx(lang = "") {
  if (_canvasContext && _canvasContext.canvas.lang !== lang) {
```

Snuffleupagus replied:
Note that the default value is `null`, from the API, so this code still needs to handle that case as well.

Snuffleupagus reviewed (changes requested):
Please also add a `text` reference test, using the first page of the affected PDF document.

And please improve the *commit message* to explain what's being changed and why, i.e. you'll want more than a single line, since it's currently difficult to understand what the patch does without looking at the code.

Aditi-1400 commented on test/integration/text_layer_spec.mjs:
```diff
@@ -290,4 +290,44 @@ describe("Text layer", () => {
+          const expectedWidth = 227;
+          const expectedWidthChrome = 242;
+          // In Chrome, an extra space character is included at the end
+          // of the text-span resulting in difference in widths
```
The difference in selection:
Chrome: 
![image](https://github.com/mozilla/pdf.js/assets/62544124/d607ae25-0b0d-4653-8a3a-c35efb92cb9f)

Firefox: 
<img width="361" alt="firefox" src="https://github.com/mozilla/pdf.js/assets/62544124/e8c6bff6-ec8c-4b07-8307-81d3a2af6ec7">

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -62,11 +62,13 @@ const MAX_TEXT_DIVS_TO_RENDER = 100000;
 
 function getCtx(lang = null) {
-  if (!_canvasContext) {
+  lang = lang || "";
+  let canvasContext = _canvasContexts.get(lang);
```
Nit:
```suggestion
  let canvasContext = _canvasContexts.get(lang ||= "");
```

Snuffleupagus reviewed (commented):
Unfortunately this will now require another rebase, sorry about that!

Aditi-1400 commented:
> Unfortunately this will now require another rebase, sorry about that!

On it!

nicolo-ribaudo commented on src/display/text_layer.js:
```diff
@@ -423,13 +423,15 @@ class TextLayer {
   static #getCtx(lang = null) {
-    if (!this.#canvasCtx) {
+    let canvasContext = this.#canvasContexts.get((lang ||= ""));
+    if (!this.#canvasContexts) {
```
```suggestion
    if (!canvasContext) {
```

Snuffleupagus commented on src/display/text_layer.js:
```diff
@@ -423,13 +423,15 @@ class TextLayer {
-
-    this.#canvasCtx?.canvas.remove();
-    this.#canvasCtx = null;
+    for (const canvasContext of this.#canvasContexts.values()) {
+      canvasContext.canvas.remove();
```
```suggestion
    for (const { canvas } of this.#canvasContexts.values()) {
      canvas.remove();
```

Snuffleupagus commented:
/botio test

Snuffleupagus reviewed (approved):
r=me, thank you!

Snuffleupagus commented:
/botio makeref
